### PR TITLE
Decompress or untar the wasm binary if it is needed when pulling it via HTTP/HTTPS

### PR DESCRIPTION
**Pull request by ingwonsong**
Decompress or/and untar the wasm binary when it is pulled via HTTP/HTTPS.
By this, we can support the compressed stream as well (even with the wrong `content-type`)

In details, when pulling bytes from HTTP/HTTPS server, if the bytes do not have proper magic number for Wasm Binary, 
tries to ungzip or untar based on the magic numbers of each fileformat.

In case of `untar`, this PR assumes that there is only one file in the same sense that OCI Image specification for Wasm binary.

Basically, it does the best effort based on the magic number of bytes.
If it fails to decompress or untar, just uses the original bytes.

Basically, the original idea came from https://github.com/istio/istio/pull/39225#issuecomment-1144207752.

### CODE REVIEW

File: pkg/wasm/httpfetcher.go [hunk 1]
```diff
@@ -15,16 +15,23 @@
...
 	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"time"
 
 	"github.com/cenkalti/backoff/v4"
 )
 
+var tarMagicNumber = []byte{0x75, 0x73, 0x74, 0x61, 0x72}
```
Comment (vikaschoudhary16) [inline pkg/wasm/httpfetcher.go:32]: just for my learning, from where you got these magic numbers?

Comment (ingwonsong) [reply pkg/wasm/httpfetcher.go:32]: You can check them various documents. (e.g., https://en.wikipedia.org/wiki/List_of_file_signatures)

**Review by GregHanson — APPROVED** [review]

**Comment by ingwonsong** [comment]
@howardjohn Do you have any thought?

**Review by howardjohn — APPROVED** [review]

**Review by zirain — APPROVED** [review]

File: pkg/wasm/httpfetcher.go [hunk 2]
```diff
@@ -15,16 +15,25 @@
...
 	"io"
+	"io/ioutil"
 	"net/http"
 	"time"
 
 	"github.com/cenkalti/backoff/v4"
 )
 
+var (
+	tarMagicNumber = []byte{0x75, 0x73, 0x74, 0x61, 0x72}
```
Comment (zirain) [inline pkg/wasm/httpfetcher.go:33]: address where these came from?

**Comment by zirain** [comment]
hold for address comments, feel free to remove it

File: pkg/wasm/httpfetcher.go (hunk 2 above)
Comment (ingwonsong) [reply pkg/wasm/httpfetcher.go:34]: Done!
